File: Algo/DynamicProgramming/TakeYouForward/practice/tuf/SlidingWindow/BinarySubarraySum.java
package SlidingWindow;

public class BinarySubarraySum {
    int[] sum = {1,0,1,0,1};
    int goal = 2;
    public static void main(String[] args) {
        new BinarySubarraySum();
    }

    BinarySubarraySum(){
        System.out.println(numSubarraysWithSum(sum, goal));
    }

    public int numSubarraysWithSum(int[] nums, int goal) {
        int i = 0, j = 0, count = 0, sum = 0, temp = 0;
        while (j<nums.length) {
            sum += nums[j];
            if (nums[j] == 1) temp = 0;
            if (sum > goal) {
                sum -= nums[i];
                i++;
            }
            while (sum == goal && i <= j) {
                temp++;
                sum -= nums[i];
                i++;
            }
            j++;
            count += temp;
        }
        return count;  
    }
}

File: Algo/DynamicProgramming/TakeYouForward/practice/practice3/slidingwindow/BinarySubarraySum.py
from typing import List


class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        sum=0
        n= len(nums)
        end=0
        myDict=dict()
        myDict[0]=1
        count=0
        while end<n:
            sum +=nums[end]
            if sum>=goal:
                count += myDict.get(sum-goal,0)
            myDict[sum]=myDict.get(sum,0)+1
            print(end,myDict,count)
            end+=1
        return count

nums = [1,0,1,0,1]
goal = 2
sol = Solution()
print(nums)
print(sol.numSubarraysWithSum(nums,goal))